src/snowflake/cli/plugins/nativeapp/artifacts.py

```diff
@@ -75,37 +60,300 @@ class NotInDeployRootError(ClickException):
+                    False,
+                )
+
+        if not match_found:
+            raise SourceNotFoundError(src)
+
+    def add(self, mapping: Union[ArtifactMapping, PathMapping]) -> None:
+        """
+        Adds an artifact mapping rule to this instance.
+        """
+        if isinstance(mapping, ArtifactMapping):
+            self._add_mapping(mapping.src, mapping.dest)
+        elif isinstance(mapping, PathMapping):
+            self._add_mapping(mapping.src, mapping.dest)
+        else:
+            raise RuntimeError(f"Unsupported mapping type: {type(mapping)}")
+
+    def _mappings_for_source(
+        self,
+        src: Path,
+        absolute: bool = False,
+        walk_directories: bool = False,
+        predicate: ArtifactPredicate = lambda src, dest: True,
+    ) -> Iterator[Tuple[Path, Path]]:
+        canonical_src = self._canonical_src(src)
+        canonical_dest = self._src_to_dest.get(canonical_src)
+        if canonical_dest is None:
+            raise SourceNotFoundError(str(src))
+
+        absolute_src = self._absolute_src(canonical_src)
+        absolute_dest = self._absolute_dest(canonical_dest)
+
+        if absolute:
+            src_for_output, dest_for_output = absolute_src, absolute_dest
+        else:
+            src_for_output, dest_for_output = canonical_src, canonical_dest
+
+        if predicate(src_for_output, dest_for_output):
+            yield src_for_output, dest_for_output
+
+        if absolute_src.is_dir() and walk_directories:
+            # both src and dest are directories, and walking directories was requested. Traverse src, and map each file
+            # to the dest directory
+            for (root, subdirs, files) in os.walk(absolute_src, followlinks=True):
+                relative_root = Path(root).relative_to(absolute_src)
+                for name in itertools.chain(subdirs, files):
+                    src_file_for_output = src_for_output / relative_root / name
+                    dest_file_for_output = dest_for_output / relative_root / name
+                    if predicate(src_file_for_output, dest_file_for_output):
+                        yield src_file_for_output, dest_file_for_output
+
+    def all_mappings(
+        self,
+        absolute: bool = False,
+        walk_directories: bool = False,
+        predicate: ArtifactPredicate = lambda src, dest: True,
+    ) -> Iterator[Tuple[Path, Path]]:
+        """
+        Yields a (src, dest) pair for each deployed artifact in the project. Each pair corresponds to a single file
+        in the project. Source directories are resolved as needed to resolve their contents.
+
+        Arguments:
+            self: this instance
+            absolute (bool): Specifies whether the yielded paths should be joined with the project or deploy roots,
+             as appropriate.
+            walk_directories (bool): Specifies whether directory to directory mappings should be expanded to
+             resolve their contained files.
```

Comment:
For what it's worth, when I looked at the usage of `all_mappings` I thought `walk_directories` would do the opposite (i.e. would directories show up in the generated pairs, or just the files?) but now I realize the question is whether or not we go through the directory boundary when it's one of the members.

This is a major nitpick, but `expand_directories` I think would be more precise.

Reply:
I used `walk` because `os.walk` was the analogy in my head, by expand is fine too.